cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/gvn/ValueNumber.qll

```diff
@@ -0,0 +1,264 @@
+        def.getResultMemoryAccess() instanceof IndirectMemoryAccess or
+        not def.hasMemoryResult()
+      )
+    )
+  }
+}
+
+/**
+ * Holds if this library knows how to assign a value number to the specified instruction, other than
+ * a `unique` value number that is never shared by multiple instructions.
+ */
+private predicate numberableInstruction(Instruction instr) {
+  instr instanceof VariableAddressInstruction or
+  instr instanceof InitializeParameterInstruction or
+  instr instanceof InitializeThisInstruction or
+  instr instanceof ConstantInstruction or
+  instr instanceof FieldAddressInstruction or
+  instr instanceof BinaryInstruction or
+  instr instanceof UnaryInstruction or
+  instr instanceof PointerArithmeticInstruction or
+  instr instanceof CongruentCopyInstruction
+}
+
+private predicate variableAddressValueNumber(VariableAddressInstruction instr, FunctionIR funcIR,
+    IRVariable var) {
+  instr.getFunctionIR() = funcIR and
+  instr.getVariable() = var
+}
+
+private predicate initializeParameterValueNumber(InitializeParameterInstruction instr,
+    FunctionIR funcIR, IRVariable var) {
+  instr.getFunctionIR() = funcIR and
+  instr.getVariable() = var
+}
+
+private predicate initializeThisValueNumber(InitializeThisInstruction instr, FunctionIR funcIR) {
+  instr.getFunctionIR() = funcIR
+}
+
+private predicate constantValueNumber(ConstantInstruction instr, FunctionIR funcIR, Type type,
+    string value) {
+  instr.getFunctionIR() = funcIR and
+  instr.getResultType() = type and
+  instr.getValue() = value
+}
+
+private predicate fieldAddressValueNumber(FieldAddressInstruction instr, FunctionIR funcIR,
+    Field field, ValueNumber objectAddress) {
+  instr.getFunctionIR() = funcIR and
+  instr.getField() = field and
+  valueNumber(instr.getObjectAddress()) = objectAddress
+}
+
+private predicate binaryValueNumber(BinaryInstruction instr, FunctionIR funcIR, Opcode opcode,
+    Type type, ValueNumber leftOperand, ValueNumber rightOperand) {
+  instr.getFunctionIR() = funcIR and
+  (not instr instanceof PointerArithmeticInstruction) and
+  instr.getOpcode() = opcode and
+  instr.getResultType() = type and
+  valueNumber(instr.getLeftOperand()) = leftOperand and
+  valueNumber(instr.getRightOperand()) = rightOperand
+}
+
+private predicate pointerArithmeticValueNumber(PointerArithmeticInstruction instr,
+    FunctionIR funcIR, Opcode opcode, Type type, int elementSize, ValueNumber leftOperand,
+    ValueNumber rightOperand) {
+  instr.getFunctionIR() = funcIR and
+  instr.getOpcode() = opcode and
+  instr.getResultType() = type and
+  instr.getElementSize() = elementSize and
+  valueNumber(instr.getLeftOperand()) = leftOperand and
+  valueNumber(instr.getRightOperand()) = rightOperand
+}
+
+private predicate unaryValueNumber(UnaryInstruction instr, FunctionIR funcIR, Opcode opcode,
+    Type type, ValueNumber operand) {
+  instr.getFunctionIR() = funcIR and
+  (not instr instanceof InheritanceConversionInstruction) and
```

Comment:
Is it impossible to value number `InheritanceConversionInstruction`s or is it just not particularly useful at the moment?

Reply:
That was just an oversight. Added, including test case.